core/src/main/scala/caliban/wrappers/Caching.scala

```diff
@@ -0,0 +1,329 @@
+  case class CachePolicy(hint: CacheHint) {
+    def merge(that: CachePolicy): CachePolicy =
+      CachePolicy(mostRestrictive(hint, that.hint))
+
+    def restrict(restricted: Option[CacheHint]): CachePolicy = restricted match {
+      case Some(h) => copy(hint = mostRestrictive(hint, h))
+      case None    => this
+    }
+
+    def toResponseValue(toHeader: CachePolicy => String) =
+      ObjectValue(
+        List(
+          "version"    -> IntValue(2),
+          "httpHeader" -> StringValue(toHeader(this))
+        )
+      )
+  }
+
+  sealed trait CacheScope
+
+  object CacheScope {
+
+    @GQLName("PRIVATE")
+    case object Private extends CacheScope {
+      override def toString: String = "PRIVATE"
+    }
+
+    @GQLName("PUBLIC")
+    case object Public extends CacheScope {
+      override def toString: String = "PUBLIC"
+    }
+
+    implicit val ord: Ord[CacheScope] = Ord.make {
+      case (Private, Private) => Ordering.Equals
+      case (Public, Public)   => Ordering.Equals
+      case (Private, Public)  => Ordering.LessThan
+      case (Public, Private)  => Ordering.GreaterThan
+    }
+
+    implicit val ordering: scala.math.Ordering[CacheScope] = ord.toScala
+
+    val _type = __Type(
+      __TypeKind.ENUM,
+      Some("CacheScope"),
+      Some("The scope of the cache"),
+      enumValues = _ =>
+        Some(
+          List(
+            __EnumValue(
+              "PRIVATE",
+              Some("The cache is private to the user"),
+              deprecationReason = None,
+              isDeprecated = false,
+              directives = None
+            ),
+            __EnumValue(
+              "PUBLIC",
+              Some("The cache is public"),
+              deprecationReason = None,
+              isDeprecated = false,
+              directives = None
+            )
+          )
+        )
+    )
+  }
+
+  case class CacheHint(
+    maxAge: Option[Duration] = None,
+    scope: Option[CacheScope] = None,
+    inheritMaxAge: Boolean = false
+  ) { self =>
+    def toHeaderString: String = {
+      val computedMaxAge = maxAge.fold(0L)(_.toSeconds)
+      val cacheable      = computedMaxAge > 0L
+      if (!cacheable) "no-store"
+      else {
+        val b = new StringBuilder()
+        b.append("max-age=")
+        b.append(computedMaxAge)
+        scope.foreach { s =>
+          b.append(", ")
+          b.append(s.toString.toLowerCase)
+        }
+        b.toString()
+      }
+    }
+  }
+
+  object CacheHint {
+    val default: CacheHint = CacheHint()
+  }
+
+  private class Default[-R1](inner: Ref[CachePolicy] => Wrapper[R1]) extends GraphQLAspect[Nothing, R1] {
+    private val _typeCache = new ConcurrentHashMap[String, Option[CacheHint]]()
+
+    def apply[R <: R1](gql: GraphQL[R]): GraphQL[R] = {
+      val wrapper = EffectfulWrapper(
+        cacheOverride.set(None) *> Ref
+          .make(CachePolicy(CacheHint.default))
+          .map(state => staticWrapper(state) |+| fieldWrapper(state) |+| inner(state))
```

Comment:
Very minor: Perhaps skip the validation wrapper for introspection queries

```suggestion
          .map(state => staticWrapper(state).skipForIntrospection |+| fieldWrapper(state) |+| inner(state))
```